Add libraries required to build UFS-Aerosols

rmontuoro opened the pull request:
This PR adds a set of new libraries required to build the UFS aerosol component based on the NASA GOCART model.

Note that the following libraries are also added for convenience to create modules used to build MAPL. These libraries, however, are not required to build UFS-Aerosols.

- ESMA_cmake, v3.4.3
- CMakeModules

rmontuoro committed:
Add libraries required to build UFS-Aerosols.

aerorahul commented on libs/build_esma_cmake.sh:
```diff
@@ -0,0 +1,40 @@
+                               || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+  fi
+else
+  prefix=${CMAKEMODULES_ROOT:-"/usr/local"}
```
This is incorrect and should be `ESMA_CMAKE_ROOT`.

rmontuoro replied:
Addressed in 2c049520a84a6938bc1724058b4e2d81c30376ae. Apologies for the multiple typos.

aerorahul commented on libs/build_gftl_shared.sh:
```diff
@@ -0,0 +1,47 @@
+                               || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+  fi
+else
+  prefix=${CMAKEMODULES_ROOT:-"/usr/local"}
```
same as above.

rmontuoro replied:
Addressed in 2c049520a84a6938bc1724058b4e2d81c30376ae

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+  module load ecbuild
+  module load gftl-shared
+  module load yafyaml
+  module load esmf
+  module load netcdf
```
this order should be reversed.

rmontuoro replied:
Addressed in 2c049520a84a6938bc1724058b4e2d81c30376ae

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+  source $MODULESHOME/init/bash
+  module load hpc-$HPC_COMPILER
+  module load hpc-$HPC_MPI
+  module load hpc-$HPC_PYTHON
```
Is MAPL needing python in the build here?

rmontuoro replied:
Yes

aerorahul replied:
where is MAPL using Python during build and installation?

This is a snippet from the `MAPL/CMakeLists.txt`
```
install (
   DIRECTORY Python
   DESTINATION lib
   USE_SOURCE_PERMISSIONS
   )
```
It is simply copying the contents of the `Python` directory to the installation location.

`hpc-$HPC_PYTHON` module is not required for building and installing MAPL.
Please remove.

rmontuoro replied:
Done in 152b70e64a501f09f4dd9eab49d7fe92d4c2e2fa.

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+                               || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+  fi
+else
+  prefix=${CMAKEMODULES_ROOT:-"/usr/local"}
```
MAPL_ROOT

rmontuoro replied:
Addressed in 2c049520a84a6938bc1724058b4e2d81c30376ae

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+  set +x
+  source $MODULESHOME/init/bash
+  module load hpc-$HPC_COMPILER
+  module load hpc-$HPC_MPI
```
This should be protected, unless MAPL needs MPI.

rmontuoro replied:
MAPL should be built with MPI. Commit faa72145badc60ae1cfed752f3ca2651756140e8 addresses this issue.

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+  export FC=$MPI_FC
+  export CC=$MPI_CC
+  export CXX=$MPI_CXX
+else
+  export FC=$SERIAL_FC
+  export CC=$SERIAL_CC
+  export CXX=$SERIAL_CXX
+fi
```
If MAPL needs MPI, this is useless and should be removed.

rmontuoro replied:
Done in faa72145badc60ae1cfed752f3ca2651756140e8

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,82 @@
+
+cmake .. \
+      -DCMAKE_INSTALL_PREFIX=$prefix \
+      -DCMAKE_MODULE_PATH=$CMAKE_MODULE_PATH \
```
See the comment in the CMakeModules and ESMA_CMake modulefiles.

aerorahul replied:
`$CMAKE_MODULE_PATH` should be replaced with `${CMAKEMODULES_ROOT}/Modules;${ESMA_CMAKE_ROOT}`

rmontuoro replied:
This is done in faa72145badc60ae1cfed752f3ca2651756140e8. Thanks for the suggestion!

aerorahul commented on libs/build_yafyaml.sh:
```diff
@@ -0,0 +1,48 @@
+                               || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+  fi
+else
+  prefix=${CMAKEMODULES_ROOT:-"/usr/local"}
```
This is repeatedly incorrect.

rmontuoro replied:
Addressed in 2c049520a84a6938bc1724058b4e2d81c30376ae

aerorahul commented on modulefiles/core/cmakemodules/cmakemodules.lua:
```diff
@@ -0,0 +1,21 @@
+whatis("Name: ".. pkgName)
+whatis("Version: " .. pkgVersion)
+whatis("Category: Software")
+whatis("Description: CMakeModules (A collection of ECMWF CMake modules)")
```
This is not from ECMWF.

rmontuoro replied:
Removed ECMWF reference in f9f91faba8371301c2248bf53303fd6783579cba

aerorahul commented on modulefiles/core/cmakemodules/cmakemodules.lua:
```diff
@@ -0,0 +1,21 @@
+
+local base = pathJoin(opt,"core",pkgName,pkgVersion)
+
+prepend_path("CMAKE_MODULE_PATH", pathJoin(base,"Modules"), ";")
```
This is not an appropriate variable to set.  The use of this variable is in cmake and cmake does not look at this environment variable.  To use this variable with cmake, one has to pass it as an command line argument, like so:
```
cmake -DCMAKE_MODULE_PATH=/path/to/append/to/default/CMAKE_MODULE_PATH
```

I suggest using the environment variable `CMAKEMODULES_ROOT` defined in this modulefile wherever necessary.

rmontuoro replied:
This setting was removed in f9f91faba8371301c2248bf53303fd6783579cba

aerorahul commented on modulefiles/core/cmakemodules/cmakemodules.lua:
```diff
@@ -0,0 +1,21 @@
+
+prepend_path("CMAKE_MODULE_PATH", pathJoin(base,"Modules"), ";")
+
+setenv("cmakemodules_ROOT", base)
```
This is fine, but use UPPERCASE as most shell env. variables are defined that way.
e.g. `$> printenv`

rmontuoro replied:
Done in f9f91faba8371301c2248bf53303fd6783579cba.

aerorahul commented on modulefiles/core/esma_cmake/esma_cmake.lua:
```diff
@@ -0,0 +1,21 @@
+
+local base = pathJoin(opt,"core",pkgName,pkgVersion)
+
+prepend_path("CMAKE_MODULE_PATH", base, ";")
```
See the comment in the CMakeModules modulefile.

rmontuoro replied:
Done in f9f91faba8371301c2248bf53303fd6783579cba.

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/mapl/mapl.lua:
```diff
@@ -0,0 +1,28 @@
+local base = pathJoin(opt,compNameVerD,mpiNameVerD,pkgName,pkgVersion)
+
+setenv("MAPL_ROOT", base)
+prepend_path("CMAKE_MODULE_PATH", pathJoin(base, "share/MAPL/cmake"), ";")
```
This should be unnecessary, if MAPL was built and packaged correctly.
If it is necessary, the use should be as described in the CMakeModules.

rmontuoro replied:
Agreed. Unfortunately this is still needed, but `${MAPL_ROOT}/share/MAPL/cmake` is now used when building the UFS weather model. `CMAKE_MODULE_PATH` is no longer set in this module (see f7aad29cbce68753b99bff1017b6e8ad15de3a12).

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/mapl/mapl.lua:
```diff
@@ -0,0 +1,28 @@
+setenv("MAPL_ROOT", base)
+prepend_path("CMAKE_MODULE_PATH", pathJoin(base, "share/MAPL/cmake"), ";")
+
+setenv("I_MPI_FABRICS", "shm:ofa")
```
Definitely not to be set in the modulefile.  
This should be set where it is used.

rmontuoro replied:
Addressed in f7aad29cbce68753b99bff1017b6e8ad15de3a12.

aerorahul commented on stack/stack_noaa.yaml:
```diff
@@ -36,7 +36,7 @@ udunits:
   build: YES
   version: 1.10.6
-  shared: NO
+  shared: YES
```
Please do not make NOAA libraries shared

rmontuoro replied:
This setting includes shared libraries without removing the static libraries. This is needed to build MAPL at this time, but I'll work with NASA to remove this requirement.

rmontuoro replied:
It turns out this requires a minor change in the MAPL build system. I've submitted an issue and will revert this change once a new MAPL version is released.

rmontuoro replied:
This issue is caused by an incomplete `NetCDF::NetCDF_Fortran` static target created by `FindNetCDF.cmake` in `CmakeModules`. This target only provides `netcdff.a`, not the full list of required libraries. See issue [#51](https://github.com/NOAA-EMC/CMakeModules/issues/51).

@aerorahul, @kgerheiser - could this issue be overcome with custom environment settings?

rmontuoro replied:
The following setting works on Hera and it does not need additional shared libraries:
```
-DNetCDF_Fortran_EXTRA_LIBRARIES="`nc-config --libs`"
```
Please let me know if this change is acceptable.

kgerheiser replied:
I think that's fine.

rmontuoro replied:
Done. Please let me know if there are additional issues in regard to CI.

aerorahul commented on stack/stack_noaa.yaml:
```diff
@@ -47,7 +47,7 @@ pnetcdf:
 netcdf:
   build: YES
-  shared: NO
+  shared: YES
```
please revert.

rmontuoro replied:
See comment above.

rmontuoro replied:
Reverted to static-only in f8afa5b128a5b3e85239b652999891fe56c7e1b4.

aerorahul commented on stack/stack_noaa.yaml:
```diff
@@ -296,3 +296,28 @@ atlas:
+cmakemodules:
+  build: YES
+  repo: NOAA-EMC
+  version: develop
```
we need to make a release, but until then, this is fine.
@kgerheiser Thoughts?

kgerheiser replied:
A release of CMakeModules? Sure we can do that.

kgerheiser replied:
@rmontuoro I made a release of CMakeModules could you update the version to `v1.2.0`?

rmontuoro replied:
Done. See 4efc88048c529a90a54691bf0661140abbe75aa4.

aerorahul commented on stack/stack_noaa.yaml:
```diff
@@ -296,3 +296,28 @@ atlas:
+
+mapl:
+  build: YES
+  repo: GEOS-ESM
```
are there multiple repos for this?
We should point to the authoritative repo.

Same applies for gftl_shared, and yafyaml.

rmontuoro replied:
No. The new entries point to authoritative repositories:
- https://github.com/Goddard-Fortran-Ecosystem/gFTL-shared
- https://github.com/Goddard-Fortran-Ecosystem/yaFyaml
- https://github.com/GEOS-ESM/MAPL

aerorahul reviewed (changes requested):
Please see inline comments.

aerorahul commented:
Also, please add the newly added components to `stack_custom.yaml` and `stack_mac.yaml` for CI.

rmontuoro committed:
Fix name of environment variable provding the installation
path if LMod is not being used.

Reverse module loading order as requested (see PR #252).

rmontuoro committed:
Require MPI and set CMAKE_MODULE_PATH according
to recommendations from PR #252.

rmontuoro committed:
Use uppercase for environment variables set in modulefiles.

Do not set CMAKE_MODULE_PATH in modulefiles.

rmontuoro committed:
Remove CMAKE_MODULE_PATH and Intel MPI settings from MAPL modulefile.

rmontuoro committed:
Add UFS-Aerosols dependencies to custom and Mac stack files.

rmontuoro commented:
> Also, please add the newly added components to `stack_custom.yaml` and `stack_mac.yaml` for CI.

Done in c75b290d15a82a2d8c5739aa9cd464946b57163c.

kgerheiser commented:
The CI is failing because MAPL can't find ESMF when built without modules. I think you need to set `ESMFMKFILE` in `build_mapl.sh`

rmontuoro commented:
> The CI is failing because MAPL can't find ESMF when built without modules. I think you need to set `ESMFMKFILE` in `build_mapl.sh`

Would it be better to set `ESMFMKFILE` in the ESMF modulefile?

kgerheiser commented:
This is when building _without_ modules. On the `else` side of `if $MODULES`.

I think `$ESMF_ROOT/lib/esmf.mk` should do the trick.

rmontuoro committed:
Set ESMFMKFILE when building MAPL without LMod.

rmontuoro commented:
> This is when building _without_ modules. On the `else` side of `if $MODULES`.
> 
> I think `$ESMF_ROOT/lib/esmf.mk` should do the trick.

Done. Thank you!

rmontuoro committed:
Add NetCDF linker settings required to overcome build issue
caused by incomplete CMake target for static NetCDF libraries.

rmontuoro committed:
Revert to static-only NetCDF and HDF5 libraries.

kgerheiser committed:
Set ecbuild module path so CMakeModules FindNetCDF takes precedence

rmontuoro committed:
Merge pull request #1 from kgerheiser/feature/gocart

Set ecbuild module path so CMakeModules FindNetCDF takes precedence

aerorahul commented on libs/build_cmakemodules.sh:
```diff
@@ -0,0 +1,40 @@
+set -eux
+
+name="cmakemodules"
+repo=${1:-${STACK_cmakemodules_repo:-"NOAA-EMC"}}
```
NOAA-EMC is the only repo from which we will install in the stack.
Please remove $repo from the `prefix` path and modulefiles.

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_esma_cmake.sh:
```diff
@@ -0,0 +1,40 @@
+set -eux
+
+name="esma_cmake"
+repo=${1:-${STACK_esma_cmake_repo:-"GEOS-ESM"}}
```
Are there multiple repos for esma_cmake?
If not, we should only  install the official one.
Please remove `$repo` from the prefix path and modulefiles.

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_esma_cmake.sh:
```diff
@@ -0,0 +1,40 @@
+mkdir -p $prefix && cp -r $software/* $prefix
+
+# generate modulefile from template
+$MODULES && update_modules core $name $repo-$id
```
should be just `$id`

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_esma_cmake.sh:
```diff
@@ -0,0 +1,40 @@
+  module list
+  set -x
+
+  prefix="${PREFIX:-"/opt/modules"}/core/$name/$repo-$id"
```
should be just `$id`

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_cmakemodules.sh:
```diff
@@ -0,0 +1,40 @@
+  module list
+  set -x
+
+  prefix="${PREFIX:-"/opt/modules"}/core/$name/$repo-$id"
```
should be just `$id`

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_cmakemodules.sh:
```diff
@@ -0,0 +1,40 @@
+mkdir -p $prefix && cp -r $software/* $prefix
+
+# generate modulefile from template
+$MODULES && update_modules core $name $repo-$id
```
should be just `$id`

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_gftl_shared.sh:
```diff
@@ -0,0 +1,47 @@
+set -eux
+
+name="gftl-shared"
+repo=${1:-${STACK_gftl_shared_repo:-"Goddard-Fortran-Ecosystem"}}
```
Same comment as above for esma-cmake and CMakeModules.

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_mapl.sh:
```diff
@@ -0,0 +1,79 @@
+set -eux
+
+name="mapl"
+repo=${1:-${STACK_mapl_repo:-"GEOS-ESM"}}
```
Here as well.

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on libs/build_yafyaml.sh:
```diff
@@ -0,0 +1,48 @@
+set -eux
+
+name="yafyaml"
+repo=${1:-${STACK_yafyaml_repo:-"Goddard-Fortran-Ecosystem"}}
```
same comment as above.

rmontuoro replied:
Done in 81e539f4a6443c4253b9e4825eeeebcd28bda884

aerorahul commented on modulefiles/compiler/compilerName/compilerVersion/gftl-shared/gftl-shared.lua:
```diff
@@ -0,0 +1,24 @@
+local base = pathJoin(opt,compNameVerD,pkgName,pkgVersion)
+
+setenv("GFTL_SHARED_ROOT", base)
+setenv("GFTL_ROOT", base)
```
The package is GFTL-Shared, what is GFTL? Is there a `non-shared` version of this library? If so, this can be misleading.

kgerheiser replied:
I didn't understand this either.

There's gftl-shared: https://github.com/Goddard-Fortran-Ecosystem/gFTL-shared

And gFTL: https://github.com/Goddard-Fortran-Ecosystem/gFTL

I think gftl-shared contains concrete containers of the most common types you would encounter like 32-bit, 64-bit (real, int, etc), while gFTL contains the templated libraries.

rmontuoro replied:
That's correct. gFTL only provides template files required by gFTL-shared, which includes gFTL as submodule.

aerorahul reviewed (changes requested):
Some more comments regarding the use of official repositories.

rmontuoro committed:
Always use authoritative repositories for CMakeModules,
ESMA_cmake, gFTL-shared, yaFyaml, and MAPL.

kgerheiser commented:
I can't figure out why the build is failing on Ubuntu. HDF5 has an error linking to pthreads, but I'm not sure why. I'm pretty sure pthreads is a built-in library. It was working previously.

rmontuoro committed:
Use CMakeModules v1.2.0.

kgerheiser commented:
Looks like everything is resolved. I am trying to figure out the CI failure, and then this can be merged.

kgerheiser commented:
Changing the OS to Ubuntu 18.04 works

kgerheiser commented on stack/stack_custom.yaml:
```diff
@@ -39,7 +39,7 @@ udunits:
 hdf5:
   build: YES
   version: 1.10.6
-  shared: YES
```
This is the line causing the CI to fail. I think this should be kept as `shared: YES`. We use `stack_NOAA.yaml` for static libraries with `shared: NO`. On user systems shared libraries are ok, but UFS likes static libraries on HPC systems.

Why it fails with Ubuntu 20.04 and GCC-9, but not Ubuntu 18.04 and GCC-7, I don't know.

rmontuoro replied:
Restored shared libraries for custom and Mac stacks (fd8237c9a467d55de02a9440c7538096bf806baa)

kgerheiser commented on stack/stack_mac.yaml:
```diff
@@ -39,7 +39,7 @@ udunits:
 hdf5:
   build: YES
   version: 1.10.6
-  shared: YES
```
Same for `stack_mac.yaml`

rmontuoro replied:
Addressed by fd8237c9a467d55de02a9440c7538096bf806baa

rmontuoro committed:
Build shared NetCDF and HDF5 libraries for custom and Mac stacks.

kgerheiser commented:
Thanks @rmontuoro. I anticipate the builds will pass now.

kgerheiser commented:
The build gets a lot further on Ubuntu now, and fails in MAPL.

```
/home/runner/work/hpc-stack/hpc-stack/mpich/pkg/mapl-v2.7.1/shared/Interp/Interp_implementation.F90:1:25:
1 | submodule (MAPL_InterpMod) Interp_implementation
|                         1
2 |   use, intrinsic :: iso_fortran_env, only: REAL32
 |                                                 2
Error: Symbol ‘real32’ at (1) conflicts with the symbol at (2)
/home/runner/work/hpc-stack/hpc-stack/mpich/pkg/mapl-v2.7.1/shared/Interp/Interp_implementation.F90:1:25:
1 | submodule (MAPL_InterpMod) Interp_implementation
|                         1
2 |   use, intrinsic :: iso_fortran_env, only: REAL32
3 |   use, intrinsic :: iso_fortran_env, only: REAL64
|                                                 2
Error: Symbol ‘real64’ at (1) conflicts with the symbol at (2)
```

Seems like a compiler error. It works on macOS which is using gfortran 9.4.0, but Ubuntu is using gfortran 9.3.0. I suspect that's the difference.

rmontuoro committed:
Do not load python module when building MAPL.

rmontuoro commented:
> /home/runner/work/hpc-stack/hpc-stack/mpich/pkg/mapl-v2.7.1/shared/Interp/Interp_implementation.F90:1:25:
> 1 | submodule (MAPL_InterpMod) Interp_implementation
> |                         1
> 2 |   use, intrinsic :: iso_fortran_env, only: REAL32
>  |                                                 2
> Error: Symbol ‘real32’ at (1) conflicts with the symbol at (2)
> /home/runner/work/hpc-stack/hpc-stack/mpich/pkg/mapl-v2.7.1/shared/Interp/Interp_implementation.F90:1:25:
> 1 | submodule (MAPL_InterpMod) Interp_implementation
> |                         1
> 2 |   use, intrinsic :: iso_fortran_env, only: REAL32
> 3 |   use, intrinsic :: iso_fortran_env, only: REAL64
> |                                                 2
> Error: Symbol ‘real64’ at (1) conflicts with the symbol at (2)

I've opened an issue with NASA.

kgerheiser commented:
I think I just need to update GCC. I found a bug on the GCC forums from this year that seems related.

climbfuji commented:
> I think I just need to update GCC. I found a bug on the GCC forums from this year that seems related.

From which version to which? We need to be able to support GNU 9.x for the moment.

kgerheiser commented:
It appears MAPL doesn't build with gfortran 9.3.0, but works with 9.4.0. I couldn't find a way to install 9.4.0 in the CI, so I'm trying 10.2.0.

climbfuji commented:
> It appears MAPL doesn't build with gfortran 9.3.0, but works with 9.4.0. I couldn't find a way to install 9.4.0 in the CI, so I'm trying 10.2.0.

Means we need a patch so that it works with 9.1.0+, otherwise it's a showstopper or at least will cause a longer delay, because we would need to update the GNU compilers on Hera and Cheyenne and possibly for the CI tests.

kgerheiser commented:
I'll investigate a little more and let you know

rmontuoro commented:
@kgerheiser - see comment in MAPL issue [#892](https://github.com/GEOS-ESM/MAPL/issues/892)

kgerheiser commented:
@climbfuji we have a fix that allows MAPL to build with gfortran 9.x

climbfuji commented:
> @climbfuji we have a fix that allows MAPL to build with gfortran 9.x

Thanks, that's great news.

rmontuoro commented:
A new MAPL release including the fix is being assembled and should be available soon.

rmontuoro committed:
Update MAPL to version 2.7.2.

rmontuoro commented:
@kgerheiser - Could you please look into the openmpi build failure? We may need to specify proper linking flags for C++ bindings.

kgerheiser commented:
I thought that might happen. I have a fix.

kgerheiser committed:
Build OpenMPI for correct C++ linking

kgerheiser commented:
I put in a PR to your branch. If you build OpenMPI yourself then the C++ library is linked in properly when using the Fortran wrapper.

rmontuoro committed:
Merge pull request #2 from kgerheiser/feature/gocart

Build OpenMPI for correct C++ linking

kgerheiser commented:
All builds pass now. @aerorahul any more comments?

aerorahul reviewed (approved):
Looks good.
Thanks @rmontuoro for addressing all the comments.
